src/OrchardCore.Modules/OrchardCore.Workflows/Trimming/Services/WorkflowTrimmingBackgroundTask.cs

```diff
@@ -0,0 +1,62 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using OrchardCore.BackgroundTasks;
+using OrchardCore.Entities;
+using OrchardCore.Modules;
+using OrchardCore.Settings;
+using OrchardCore.Workflows.Trimming.Models;
+
+namespace OrchardCore.Workflows.Trimming.Services;
+
+[BackgroundTask(
+    Schedule = "0 0 * * *",
+    Title = "Workflow Trimming Background Task",
+    Description = "Regularly deletes old workflow instances.",
+    LockTimeout = 3_000,
+    LockExpiration = 30_000
+)]
+public class WorkflowTrimmingBackgroundTask : IBackgroundTask
+{
+    public async Task DoWorkAsync(IServiceProvider serviceProvider, CancellationToken cancellationToken)
+    {
+        var siteService = serviceProvider.GetRequiredService<ISiteService>();
+
+        var workflowTrimmingSettings = await siteService.GetSettingsAsync<WorkflowTrimmingSettings>();
+        if (workflowTrimmingSettings.Disabled)
+        {
+            return;
+        }
+
+        var logger = serviceProvider.GetRequiredService<ILogger<WorkflowTrimmingBackgroundTask>>();
+
+        try
+        {
+            var clock = serviceProvider.GetRequiredService<IClock>();
+            var workflowTrimmingManager = serviceProvider.GetRequiredService<IWorkflowTrimmingManager>();
+            var batchSize = serviceProvider.GetRequiredService<IOptions<WorkflowTrimmingOptions>>().Value.BatchSize;
+
+            logger.LogDebug("Starting trimming Workflow instances.");
+            var trimmedCount = await workflowTrimmingManager.TrimWorkflowInstancesAsync(
+                TimeSpan.FromDays(workflowTrimmingSettings.RetentionDays),
+                batchSize
+            );
+            logger.LogDebug("Trimmed {TrimmedCount} workflow instances.", trimmedCount);
+
+            var siteSettings = await siteService.LoadSiteSettingsAsync();
+            siteSettings.Alter<WorkflowTrimmingSettings>(settings =>
```

Comment:
This should not be part of the settings. This may cause concurrency issue since the document is singleton and could be updated during run time. 

Probably a different document other a different method.